r1cs-std/src/groups/mod.rs

```diff
@@ -111,6 +111,21 @@ pub trait GroupGadget<G: Group, ConstraintF: Field>:
         Ok(())
     }
 
+    fn precomputed_base_scalar_mul_3_bit_with_conditional_negation<'a, CS, T, I, B>(
```

Comment:
From what I understand, the only difference between `precomputed_base_scalar_mul_3_bit_with_conditional_negation` and `precomputed_base_multiscalar_mul` is that the former requires an additional `segment_size` argument. Is there a way to avoid duplication in the interface itself?

Couple of options:
1) Have just one generic `precomputed_base_multiscalar_mul` whose signature is the same as `precomputed_base_scalar_mul_3_bit_with_conditional_negation`, but whose default implementation is the same as the current `precomputed_base_multiscalar_mul`. Then, for `TEGadget`, we just use the current implementation of `precomputed_base_scalar_mul_3_bit_with_conditional_negation` as the impl.
2) Investigate whether we actually need the `segment_size` argument, or whether it can be a fixed constant derivable from things like the scalar field size? I understand that, as used in the BHP hash, `segment_size` is just the window-size; is this value expected to differ across instantiations, or will it remain the same?

Reply:
That might be problematic - precomputed_base_scalar_mul_3_bit_with_conditional_negation doesn't have the same goal as precomputed_base_multiscalar_mul - it's more efficient, though it maps ([b1, b2, b3], G) to [-4,...,-1,1,...4]*G rather than [0,...,7]*G as precomputed_base_multiscalar_mul does.

I think we might be able to get rid of segment_size though.

Reply:
Not 100% sure how to cleanly get rid of segment size yet. Let me know if you got ideas.

Reply:
The reason it's not direct is because it's not directly a property of the group when to move on to the next segment. The maximum is a property though. The question is whether to always use the maximum.

Reply:
Re: merging the two, then maybe it might be worthwhile to rename the method to something different? To avoid confusion with normal multi-scalar multiplication. Do you have any ideas? (`precomputed_base_scalar_mul_3_bit_with_conditional_negation` is also a bit wordy)

If we rename it, then maybe the segment_size question is also moot, if we make it clearer what it means?. Another approach would be to potentially derive `segment_size` from the input length + group size? Do you have any references I could look at that talk about the derivation of segment_size?

Reply:
That's a good idea.

The derivation of the maximum segment size is both implemented in https://github.com/celo-org/zexe/blob/052e66bb6153383f164aba8a3847f0e13bb2e903/crypto-primitives/src/crh/bowe_hopwood/mod.rs#L57 and described in the Zcash spec (https://github.com/zcash/zips/blob/master/protocol/protocol.pdf) in page 56.

If we always use the maximum one, then no doubt we can derive it. I think we might want to do that, there shouldn't be any real advantage to doing otherwise.

Reply:
I'll leave the decision regarding `segment_size` up to you! Re: name suggestions, maybe something based on `signed_digits`? Eg:

* `precomputed_base_signed_digit_scalar_mul`, with a comment explaining that it works only for 3-bit signed digits?
* `precomputed_base_3_bit_signed_digit_scalar_mul`